### PR TITLE
#SEcKZBFF Register drivers & operators

### PR DESCRIPTION
**Pull request by Ally4**
#### What does this PR do?

- This PR is for the registration of the drivers and the operators

#### Description of Task to be completed?

- Creation of the registration controller

- Creation of the registration router

- Creation of the password utilities

- Creation of the validations of the controller

- Installation of packages

#### How should this be manually tested?

- By cloning this repository

- Running `npm install`

- Create database users and tests

- Create an account in SendGrid 

- Create an api key from that account

- Create a .env file in the main directory

- Fill it with the adequate values referencing form the .env.example

- Running `npm run dev`

- By sending the request `http://localhost:3020/registration` in postman with the post method

- By filling the fields of `email, firstname and lastname` in the body of the postman

#### The swagger documentation!.

- This is the documentation link: `https://eagles-phant-ft-registe-qzixtx.herokuapp.com/api-docs/#/User/post_registration`

#### Any background context you want to provide?

- installation of the packages

#### What are the relevant Trello stories?

- `https://trello.com/c/SEcKZBFF/10-register-drivers-operators`

#### Screen shots?
Screen Shot 2020-10-08 at 07 33 53

Screen Shot 2020-10-08 at 16 33 05

Screen Shot 2020-10-08 at 16 33 18

Screen Shot 2020-10-08 at 16 33 57

### CODE REVIEW

**Review by TuyizeeAnastase — COMMENTED** [review]
you labeled as Help wanted ? where?

**Review by Ntwali-Josue — APPROVED** [review]

**Review by Mugema123 — APPROVED** [review]

**Review by TuyizeeAnastase — APPROVED** [review]

File: .travis.yml [hunk 1]
```diff
@@ -1,4 +1,23 @@
-language: node_js
+services:
+  - postgresql
+before_script:
+  - psql -c 'CREATE DATABASE test;' -U postgres
+  - psql -c 'CREATE DATABASE users;' -U postgres
```
Comment (blaise82) [inline .travis.yml:5]: why do you create two databases?

File: .travis.yml [hunk 2]
```diff
@@ -1,4 +1,23 @@
...
+sudo: false
 
+cache:
+  directories:
+    - 'node_modules'  
+  
+language: node_js
 node_js:
- - 14.8.0
+  - "12.13.0"
```
Comment (blaise82) [inline .travis.yml]: put this on the top
and leave the version to just 12

File: package.json [hunk 3]
```diff
@@ -3,7 +3,7 @@
   "version": "1.0.0",
   "main": "app.js",
   "scripts": {
-    "test": "cross-env NODE_ENV=test nyc --reporter=html --reporter=text --all=true mocha --require @babel/polyfill --require @babel/register src/test/*.js --timeout 2000000 --exit",
+    "test": "cross-env NODE_ENV=test npm run migrate:reset && cross-env NODE_ENV=test npm run seed && cross-env NODE_ENV=test nyc --reporter=html --reporter=text --all=true mocha --require @babel/polyfill --require @babel/register src/test/*.js --timeout 5000000 --exit",
```
Comment (blaise82) [inline package.json]: create one DB migration script for tests

File: package.json [hunk 4]
```diff
@@ -19,11 +19,12 @@
...
       "src/test",
-      "src/database/models",
-      "src/database/seeders",
-      "coverage"
+      "coverage",
+      "src/middleware",
+      "src/validation",
+      "src/dummydata"
     ]
   },
```
Comment (blaise82) [inline package.json]: don't exclude anything

File: src/controllers/registration.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+      const {email, firstname, lastname, dateofbirth, gender, address } = req.body;
+      const inSystem = await Users.findOne({ where: { email: req.body.email } });
+      if (inSystem) { return res.status(409).json({ status: 409, message: 'The email is already in the system' });}
+      const newUser = await Users.create({email, firstname, lastname, passwordhashed, dateofbirth, gender, address,});
+      const payload = { email, firstname };const token = jwt.sign(payload, secret);
+      const message = { to: email, from: 'eaglesphantom1@gmail.com',subject: 'testing sendgrid', text: `this is the password ${password}`, html: `this is the password ${password}` };
+      mail.send(message);
+      return res.status(201).json({status: 201, message: 'user created successfully', data: { token } });
+    } catch (error) { return res.status(500).json({ status: 500, message: error.message }); } } }
+export default registration;
```
Comment (blaise82) [inline src/controllers/registration.js]: use express or join validation

File: src/controllers/registration.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+import signup from '../validations/signup';
+import { password, passwordhashed } from '../utilities/password';
+dotenv.config();
+mail.setApiKey(process.env.SENDGRID);
+const { Users } = models;
+const secret = process.env.SECRET_KEY || 'thisisthenewkeygh';
+class registration {
+  static async signup(req, res) {
+    try { const { error } = signup.validate(req.body);
+      if (error) { return res.status(400).json({ status: 400, message: error.details[0].message.replace(/"/g, '') }); }
```
Comment (blaise82) [inline src/controllers/registration.js]: this should be in the joi file

File: src/database/migrations/20201007150915-create-users.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
+module.exports = {
+  up: async (queryInterface, Sequelize) => {
+    await queryInterface.createTable('Users', {
+      id: { allowNull: false,
+        autoIncrement: true,
+        primaryKey: true,
+        type: Sequelize.INTEGER,
```
Comment (blaise82) [inline src/database/migrations/20201007150915-create-users.js]: change the id to uuid not an integer(autoIncrement)

File: src/database/models/users.js [hunk 8]
```diff
@@ -0,0 +1,19 @@
+import dotenv from 'dotenv';
+
+dotenv.config();
+export default (sequelize, DataTypes) => {
+  const Users = sequelize.define('Users', {
+    firstname: DataTypes.STRING,
+    lastname: DataTypes.STRING,
+    email: DataTypes.STRING,
+    passwordhashed: DataTypes.STRING,
```
Comment (blaise82) [inline src/database/models/users.js]: just call it password

**Review by blaise82 — APPROVED** [review]

File: .travis.yml (hunk 1 above)
Comment (Ally4) [reply .travis.yml:5]: let me fix it

File: .travis.yml (hunk 2 above)
Comment (Ally4) [reply .travis.yml]: node_js: "12", done

File: package.json (hunk 4 above)
Comment (Ally4) [reply package.json]: all the excludes removed

File: src/controllers/registration.js (hunk 5 above)
Comment (Ally4) [reply src/controllers/registration.js]: It has the joi validation

File: src/database/migrations/20201007150915-create-users.js (hunk 7 above)
Comment (Ally4) [reply src/database/migrations/20201007150915-create-users.js]: changed to UUID

File: src/database/models/users.js (hunk 8 above)
Comment (Ally4) [reply src/database/models/users.js]: changed to password!.

File: src/controllers/registration.js (hunk 6 above)
Comment (Ally4) [reply src/controllers/registration.js]: Done

File: src/database/migrations/20201007150915-create-users.js (hunk 7 above)
Comment (Ally4) [reply src/database/migrations/20201007150915-create-users.js]: Done

**Pull request merged**